### PR TITLE
Charlesmchen/quoted replies5

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
This PR is now ready for review.

PTAL @michaelkirk

### CODE REVIEW

**Comment by charlesmchen-signal** [comment]
This PR is now ready for review.

PTAL @michaelkirk

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+typedef NS_ENUM(NSUInteger, OWSMessageGestureLocation) {
+    // Message text, etc.
+    OWSMessageGestureLocation_Default,
+    OWSMessageGestureLocation_OversizeText,
+    OWSMessageGestureLocation_Media,
+    OWSMessageGestureLocation_QuotedReply,
+};
+
+@interface OWSMessageBubbleView : UIView
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.h]: Pulled out `OWSMessageBubbleView` that we can use in 

* conversation view cells
* message detail view.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+@property (nonatomic, nullable, readonly) UIView *lastBodyMediaView;
+
+- (instancetype)init NS_UNAVAILABLE;
+
+- (instancetype)initWithFrame:(CGRect)frame NS_DESIGNATED_INITIALIZER;
+
+- (instancetype)initWithCoder:(NSCoder *)coder NS_UNAVAILABLE;
+
+- (void)configureViews;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.h]: This view can be used for rendering or measurement. This method only applies to rendering.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    // Message text, etc.
+    OWSMessageGestureLocation_Default,
+    OWSMessageGestureLocation_OversizeText,
+    OWSMessageGestureLocation_Media,
+    OWSMessageGestureLocation_QuotedReply,
+};
+
+@interface OWSMessageBubbleView : UIView
+
+@property (nonatomic, nullable) ConversationViewItem *viewItem;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.h]: All properties need to be set before using it for rendering or measurement.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 4]
```diff
@@ -0,0 +1,1083 @@
...
+// Should do nothing if view is already loaded.
+@property (nonatomic, nullable) dispatch_block_t loadCellContentBlock;
+// Should unload all expensive view contents (images, etc.).
+@property (nonatomic, nullable) dispatch_block_t unloadCellContentBlock;
+
+@property (nonatomic, nullable) NSMutableArray<NSLayoutConstraint *> *viewConstraints;
+
+@end
+
+@implementation OWSMessageBubbleView
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: Nearly all of the logic here was pulled out, unmodified from the message cell.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 5]
```diff
@@ -0,0 +1,1083 @@
...
+    [self.lastBodyMediaView removeFromSuperview];
+    self.lastBodyMediaView = nil;
+
+    [self.lastQuotedMessageView removeFromSuperview];
+    self.lastQuotedMessageView = nil;
+}
+
+#pragma mark - Gestures
+
+- (OWSMessageGestureLocation)gestureLocationForLocation:(CGPoint)locationInMessageBubble
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: This method was added: it identifies which message content a given gesture applies to.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m [hunk 6]
```diff
@@ -1435,70 +483,74 @@ - (void)handleTextTapGesture:(UITapGestureRecognizer *)sender
         }
     }
 
-    if (self.hasTapForMore) {
-        [self.delegate didTapTruncatedTextMessage:self.viewItem];
-        return;
+    CGPoint locationInMessageBubble = [sender locationInView:self.messageBubbleView];
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Thanks to `gestureLocationForLocation` let us significantly streamline this logic.  I think it's more readable too.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 7]
```diff
@@ -0,0 +1,1083 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface OWSMessageBubbleView ()
+
+@property (nonatomic) OWSBubbleView *bubbleView;
+
+@property (nonatomic) OWSMessageTextView *bodyTextView;
+
+@property (nonatomic, nullable) UIView *lastQuotedMessageView;
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: What is the significance of "last" in this name again?

Is there a `firstQuotedMessageView` or any other `quotedMessageViews`?

Same question for `lastBodyMediaView`

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m (hunk 6 above)
Comment (michaelkirk-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: 👍

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 7 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: I've renamed both. Sorry for the confusion.

**Pull request merged**